Add optimization blacklist as second arg to optimizeForMobile method.

xcheng16 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37462 Add optimization blacklist as second arg to optimizeForMobile method.**

Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22096029](https://our.internmc.facebook.com/intern/diff/D22096029)

xcheng16 committed:
Add whitelist optimization dict as second arg to optimizeForMobile method.
Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

AshkanAliabadi commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
-  cloned_module = freeze_module(cloned_module);
-  FoldPrePackingOps(cloned_module);
+
+  auto bn_it =
```
I would personally mark these as const auto (constness is not implied otherwise here) to have the compiler help me in situations where I accidentally modify the variable, like for instance `if (bn_it = whitelist_optimizers.end())`.

AshkanAliabadi commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
+    cloned_module = freeze_module(cloned_module);
+    FoldPrePackingOps(cloned_module);
+  }
+
```
This pattern is fine for two or three use cases but when and if the list of options grow we want to switch to something like this:

```
// Psudo code - most probably won't compile without modifications.
struct OpCode2Op {
  MobileOptimizerType opcdoe;
  std::function< script::Module (const script::Module&) > op;
};

const std::vector<OpCode2Op> opcodes2ops = {
  // OP 1
  {
     MobileOptimizerType::FOLD_CONV_BATCH_NORM,
     [](const script::Module& module) {
           return FoldConvBatchNorm2d(cloned_module);
     },
  },
  // OP 2
  {
    ...
  },
  // OP 3 and so on
  ...
};

auto cloned_module = m.clone();
cloned_module.eval();

foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```

kimishpatel replied:
I like this suggestion. We can have a `const static` instance of OpCode2Op, or call it OptimizationTypes as Ashkan showed?
```
struct OptimizationType {
  MobileOptimizerType type;
  std::function< script::Module (const script::Module&) > apply;
};
```
Also we may need to run optimization in particular order. Meaning if you run batch norm fused after prepacking to insert xnnpack conv/linear then bn folding wont work. So we need a way to capture that as well. std::vector kind of works for it, but it does not capture the intent.

xcheng16 replied:
Thanks for the suggestion, It looks good. Let me create a struct and attach a function for each pass. @kimishpatel Do we need to cover running optimization in particular order here?

kimishpatel replied:
Yes the order matters. Therefore I am not sure if doing thing is necessarily good:
```
foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```
@AshkanAliabadi, what do you suggest? I am of the opinion that we shouldn't just loop over the vector. It gets rid of the order information. I suggest we do something like:
```
assuming optimization_blacklist is a set
if (!optimization_blacklist.count(ConvBNFusion)) {
optimization_types.find(ConvBNFusion)->apply();
}
```
Or we can just keep it simple to be honest. 
In fact, while I like the idea, lets do it when we actually need to do it. Right now we dont have that many so we can keep it simple.
```
if (!optimization_blacklist.count(ConvBNFusion)) {
FoldConvBatchNorm2d(module)
}
```

xcheng16 replied:
I think not looping the list and just checking if it is in blacklist before executing if order matters. @AshkanAliabadi are you good with it?

AshkanAliabadi replied:
Yes definitely.

Generally speaking anything I say is just a suggestion so please feel free to ignore.  I believe the person writing the code or developing a feature has the best context and understanding of what's going on so please do as what you think makes best sense.  Thanks again.

AshkanAliabadi commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
-c10::optional<script::Module> optimizeForMobile(const script::Module& m) {
+script::Module optimizeForMobile(
+    const script::Module& m,
+    const std::map<MobileOptimizerType, bool>& whitelist_optimizers) {
```
Also do we need a bool?  It could as well be const std::vector<MobileOptimizerType> & blacklist.  If it's in the blacklist, then skip, otherwise apply.

kimishpatel replied:
Maybe `unordered_set` if possible instead of map?

xcheng16 replied:
The initial idea was by default we run through all the optimization pass, but also let customer to customize it by passing a map where they can assign the value for each pass true/pass. Both of you are right, in that case we could have just created a blacklist vector to make it simpler.

AshkanAliabadi reviewed (commented):
Thanks a lot for working on this Xingying.

kimishpatel commented on test/test_mobile_optimizer.py:
```diff
@@ -80,6 +94,41 @@ def forward(self, x):
+        FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 1, exactly=True) \
+                   .run(str(get_forward_graph(bn_fold_scripted_module._c)))
+        bn_input = torch.rand(1, 1, 6, 6)
+        torch.testing.assert_allclose(bn_scripted_module(bn_input), bn_fold_scripted_module(bn_input), rtol=1e-2, atol=1e-3)
```
Can you add test case to check when both optimizations are turned off?

kimishpatel commented on test/test_mobile_optimizer.py:
```diff
@@ -80,6 +94,41 @@ def forward(self, x):
+        }
+        bn_fold_scripted_module = optimize_for_mobile(bn_scripted_module, whitelist_optimizer_dict_with_bn_fold)
+        self.assertEqual(len(torch.jit.export_opnames(bn_fold_scripted_module)), 1)
+        FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 1, exactly=True) \
```
It will be good to check that no prepacked ops are there. Plus not batch_norm.

kimishpatel reviewed (changes requested):
Looks great. Left a few comments.

xcheng16 committed:
Update on "Add whitelist optimization dict as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add whitelist optimization dict as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add whitelist optimization dict as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add whitelist optimization dict as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add optimization blacklist as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/utils/mobile_optimizer.py:
```diff
@@ -4,26 +4,33 @@
     Args:
         script_module: An instance of torch script module with type of ScriptModule
-
+        whitelist_optimizer_dict: A dictionary with key type of MobileOptimizerType and value type of bool.
```
Fix the comment? It is optimization_blacklist and not whitelist.

kimishpatel reviewed (approved):
Looks good. Please fix the comment before landing.

xcheng16 committed:
Update on "Add optimization blacklist as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add optimization blacklist as second arg to optimizeForMobile method."


Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22096029](https://our.internmc.facebook.com/intern/diff/D22096029)

[ghstack-poisoned]
